ledger/ledger-api-auth/src/main/scala/com/digitalasset/ledger/api/auth/OngoingAuthorizationObserver.scala

```diff
@@ -29,12 +63,85 @@ private[auth] final class OngoingAuthorizationObserver[A](
   override def setMessageCompression(b: Boolean): Unit = observer.setMessageCompression(b)
 
   override def onNext(v: A): Unit =
-    authorized(claims) match {
+    authorize match {
+      case _ if aborted.get() => ()
       case Right(_) => observer.onNext(v)
-      case Left(authorizationError) => observer.onError(throwOnFailure(authorizationError))
+      case Left(authorizationError: AuthorizationError) =>
+        val e: StatusRuntimeException =
+          errorFactories.permissionDenied(authorizationError.reason)(errorLogger)
+        observer.onError(e)
     }
 
   override def onError(throwable: Throwable): Unit = observer.onError(throwable)
 
   override def onCompleted(): Unit = observer.onCompleted()
+
+  private def authorize: Either[AuthorizationError, Unit] = {
+    val now = nowF()
+    for {
+      _ <- originalClaims.notExpired(now)
+      _ = validateClaimsResolvedFromUser(now)
+    } yield {
+      ()
+    }
+  }
```

Comment:
Inline the two calls into the `onNext` case distinction above for clarity.

Reply:
Ok